pkg/volume/flexvolume/flexvolume.go

```diff
@@ -241,76 +245,142 @@ func (f flexVolumeMounter) GetAttributes() volume.Attributes {
 type flexVolumeManager interface {
 	// Attaches the disk to the kubelet's host machine.
-	attach(mounter *flexVolumeMounter) (string, error)
+	attach(mounter *flexVolumeMounter, volOptions map[string]string) (string, error)
 	// Detaches the disk from the kubelet's host machine.
 	detach(unmounter *flexVolumeUnmounter, dir string) error
 	// Mounts the disk on the Kubelet's host machine.
-	mount(mounter *flexVolumeMounter, mnt, dir string) error
+	mount(mounter *flexVolumeMounter, mnt, dir string, volOptions map[string]string) error
 	// Unmounts the disk from the Kubelet's host machine.
-	unmount(unounter *flexVolumeUnmounter, dir string) error
+	unmount(unmounter *flexVolumeUnmounter, dir string) error
 }
 
 // SetUpAt creates new directory.
 func (f *flexVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
-
-	notmnt, err := f.blockDeviceMounter.IsLikelyNotMountPoint(dir)
-	if err != nil && !os.IsNotExist(err) {
-		glog.Errorf("Cannot validate mountpoint: %s", dir)
+	err := f.Attach(f.spec, "")
+	if err != nil {
 		return err
 	}
-	if !notmnt {
-		return nil
+	devicePath, err := f.WaitForAttach(f.spec, 1*time.Hour)
+	if err != nil {
+		return err
 	}
+	deviceMountPath := f.GetDeviceMountPath(f.spec)
+	return f.MountDevice(f.spec, devicePath, deviceMountPath, f.mounter)
+}
+
+type attachResult struct {
+	device string
+	err    error
+}
 
-	if f.options == nil {
-		f.options = make(map[string]string)
+func getVolumeSource(spec *volume.Spec) *api.FlexVolumeSource {
+	if spec.Volume != nil {
+		return spec.Volume.FlexVolume
+	}
+	return spec.PersistentVolume.Spec.FlexVolume
+}
+
+func (f *flexVolumeMounter) getVolumeOptions(spec *volume.Spec) (map[string]string, error) {
+	volumeSource := getVolumeSource(spec)
+	options := make(map[string]string)
+	for name, value := range volumeSource.Options {
+		options[name] = value
 	}
 
-	f.options[optionFSType] = f.fsType
+	options[optionFSType] = volumeSource.FSType
 
 	// Read write mount options.
-	if f.readOnly {
-		f.options[optionReadWrite] = "ro"
+	if spec.ReadOnly {
+		options[optionReadWrite] = "ro"
 	} else {
-		f.options[optionReadWrite] = "rw"
+		options[optionReadWrite] = "rw"
 	}
 
 	// Extract secret and pass it as options.
-	for name, secret := range f.secrets {
-		f.options[optionKeySecret+"/"+name] = secret
+	secrets, err := getSecrets(spec, spec.PodNamespace, f.plugin.host)
+	if err != nil {
+		return nil, err
 	}
+	for name, secret := range secrets {
+		options[optionKeySecret+"/"+name] = secret
+	}
+	return options, nil
+}
 
-	device, err := f.manager.attach(f)
+func (f *flexVolumeMounter) Attach(spec *volume.Spec, hostName string) error {
+	options, err := f.getVolumeOptions(spec)
 	if err != nil {
-		if !isCmdNotSupportedErr(err) {
-			glog.Errorf("Failed to attach volume: %s", f.volName)
-			return err
+		return err
+	}
+
+	if f.attachResultC == nil {
+		f.attachResultC = make(chan attachResult, 1)
+	}
+	go func() {
+		device, err := f.manager.attach(f, options)
```

Comment:
Yes that's really the big point. The goroutine leak would also be process leak here I think :-) This also requires that the Attach and WaitForAttach must be called on the same host/process, which is the case now (calling SetUpAt) but may not be true after the move to the plugin. Given how for instance the GCE volume plugin works, I could try to get the exec.Cmd object out of attach and watch it in WaitForAttach instead of using a chan+goroutine?